sdk/servicebus/service-bus/src/core/messageReceiver.ts

```diff
@@ -1138,3 +1145,81 @@ export class MessageReceiver extends LinkEntity {
     return result;
   }
 }
+
+/**
+ * Wraps the receiver with some higher level operations for managing state
+ * like credits, draining, etc...
+ *
+ * @internal
+ * @ignore
+ */
+export class ReceiverHelper {
+  private _stopReceivingMessages: boolean = false;
+
+  constructor(private _getCurrentReceiver: () => Receiver | undefined) {}
+
+  /**
+   * Adds credits to the receiver, respecting any state that
+   * indicates the receiver is closed or should not continue
+   * to receive more messages.
+   *
+   * @param credits Number of credits to add.
+   */
+  public addCredit(credits: number): boolean {
+    const receiver = this._getCurrentReceiver();
+
+    if (this._stopReceivingMessages || receiver == null) {
+      return false;
+    }
+
+    receiver.addCredit(credits);
+    return true;
+  }
+
+  /**
+   * Prevents us from receiving any further messages.
+   */
+  public async stopReceivingMessages(): Promise<void> {
+    const receiver = this._getCurrentReceiver();
+
+    if (receiver == null) {
+      return;
+    }
+
+    log.receiver(
+      `[${receiver.name}] User has requested to stop receiving new messages, attempting to drain the credits.`
+    );
+    this._stopReceivingMessages = true;
+
+    return this.drain();
```

Comment:
Are there any problems if you call `stopReceivingMessages()` more than once? Imagining if a user called `close()` multiple times for some reason.

Reply:
I can add a test for it but I don't think so unless you're not allowed to drain a receiver twice.

Reply:
I’m curious if the service will ever returned the drained event if it already drained.

Reply:
Just tried it. As somewhat expected it basically functions as a no-op (nothing to drain).

We do a drain after each receiveBatch call if we don't fetch all your expected messages. So we have a workflow similar to this already.

Reply:
Cool, so the service does return a flow frame indicating that the link was drained even if there were no credits on the link? Works for me then!